app/abci.go

```diff
@@ -0,0 +1,160 @@
+package app
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lazyledger/lazyledger-app/x/lazyledgerapp/types"
+	abci "github.com/lazyledger/lazyledger-core/abci/types"
+	core "github.com/lazyledger/lazyledger-core/proto/tendermint/types"
+)
+
+// This file should contain all of the althered ABCI methods
+
+// PreprocessTxs fullfills the lazyledger-core version of the ACBI interface, by
+// performing basic validation for the incoming txs, and by cleanly separating
+// share messages from transactions
+// todo(evan): refactor out a for loop.
+func (app *App) PreprocessTxs(txs abci.RequestPreprocessTxs) abci.ResponsePreprocessTxs {
+	squareSize := app.SquareSize()
+	shareCounter := uint64(0)
+	var shareMsgs []*core.Message
+	var processedTxs [][]byte
+	for _, rawTx := range txs.Txs {
+		// decode the Tx
+		tx, err := app.txConfig.TxDecoder()(rawTx)
+		if err != nil {
+			continue
+		}
+
+		// don't do anything if the transaction doesn't contain a PayForMessage sdk.Msg
+		if !hasWirePayForMessage(tx) {
+			processedTxs = append(processedTxs, rawTx)
+			continue
+		}
+
+		// only support transactions that contain a single sdk.Msg for now
+		if len(tx.GetMsgs()) != 1 {
+			continue
+		}
+
+		msg := tx.GetMsgs()[0]
+
+		// run basic validation on the transaction
+		err = tx.ValidateBasic()
+		if err != nil {
+			continue
+		}
+
+		// process and validate the message
+		coreMsg, signedTx, err := app.processMsg(msg)
+		if err != nil {
+			continue
+		}
+
+		// execute the tx in runTxModeDeliver mode (3)
+		// execution includes all validation checks burning fees
+		// currently, no fees are burned
+		_, _, err = app.BaseApp.TxRunner()(3, rawTx)
+		if err != nil {
+
+			continue
+		}
+
+		// increment the share counter by the number of shares taken by the message
+		sharesTaken := uint64(len(coreMsg.Data) / 256)
+		shareCounter += sharesTaken
+
+		// if there are too many shares stop processing and return the transactions
+		// this is related to issues 67 and 77 of lazyledger-core
+		if shareCounter > squareSize {
+			break
+		}
+
+		// encode the processed tx
+		rawProcessedTx, err := app.appCodec.MarshalBinaryBare(signedTx)
+		if err != nil {
+			continue
+		}
+
+		// add the message and tx to the output
+		shareMsgs = append(shareMsgs, &coreMsg)
+		processedTxs = append(processedTxs, rawProcessedTx)
+	}
+
+	// sort messages lexographically
+	sort.Slice(shareMsgs, func(i, j int) bool {
+		return bytes.Compare(shareMsgs[i].NamespaceId, shareMsgs[j].NamespaceId) < 0
+	})
+
+	return abci.ResponsePreprocessTxs{
+		Txs:      processedTxs,
+		Messages: &core.Messages{MessagesList: shareMsgs},
+	}
+}
```

Comment:
Note that in PreProcess the app also needs to track the (original / mempool) Tx that were processed, s.t. on re-check  it can tell ll-core to remove them from the mempool. 

This has the caveat that currently only the proposer will call preprocess: the app needs to know if a proposed block didn't make it through consensus (e.g. because of timeouts) and other nodes to be able to evict their mempools too...
@adlerjohn you and I need to think about, too.

Reply:
Not necessarily? Conflicting transactions (which includes duplicates in the mempool) between the block and the mempool are removed with CheckTx, so PreProcess doesn't technically need to do that?

Reply:
Looks like I'm confused about this again, then 🤔  
How does the app know which (mempool) Tx made it into the block if the app doesn't track this?

Reply:
Leaving this here for reference (quoting @adlerjohn): 
> You don't need to know which txs from the mempool were included in a block, you only need to know which mempool transactions conflict with the ones in a block.
> So running CheckTx of all txs in the mempool after you commit a block is sufficient.
> If the tx in the mempool conflicts. It'll have the same nonce and sender.

So my concern was the mapping between mempool tx <-> block tx but as mentioned by John that mapping is essentially there via nonce and sender.